// src/controllers/auth.controller.ts
import { Request, Response } from "express";
import bcrypt from "bcrypt";
import jwt from "jsonwebtoken";
import User from "src/models/user.model";

const JWT_SECRET = process.env.JWT_SECRET || "SportOn_6667777";

export const signin = async (req: Request, res: Response): Promise<void> => {
    try {
        const {email, password} = req.body;

        //Check if user Exited
        const user = await User.findOne({email})
        if (!user) {
            res.status(400).json({Message: "Invalid Credentials, Email not found"});
            return;
        }

        //Validate Password
        const isMatch = await bcrypt.compare(password, user.password);
        if (!isMatch) {
            res.status(400).json({Message: "Invalid Credentials, wrong password"});
            return;
        }

        //Generate JWT (JSON Web Token)
        const token = jwt.sign({id: user._id, email: user.email}, JWT_SECRET, {
            expiresIn: "1d"
        })

        res.json({
            token,
            user: {
                id: user._id,
                name: user.name,
                email: user.email,
            }
        })
    }   catch (error) {
        console.error("Signin Error : ", error)
        res.status(500).json({Message: "Internal Server Error"});
    };
};

export const initiateAdmin = async (req: Request, res: Response): Promise<void> => {
    try {
        const {email, password, name} = req.body;

        //password : Admin456852
        //password : 85245852

        //Check if user data/entry is exited

        const count = await User.countDocuments({});
        if (count > 0) {
            res.status(400).json({
                Message: "we can only have 1 admin user, if your want to crate new admin user, please delete the user manually from the databased"
            })
            return;
        }

        const salt = await bcrypt.genSalt(10);
        const hashedPassword = await bcrypt.hash(password, salt);

        const newUser = new User ({
            email,
            password: hashedPassword,
            name
        })

        await newUser.save();
        
        //res 201 is success
        res.status(201).json({Message: "Successfully!"});
    } catch (error) {
        console.error("Initiate new admin user error : ", error);
        res.status(500).json({Message: "Internal Server User"});
    }
};
